### PR TITLE
Add log history filters

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: inc/html.class.php [hunk 1]
```diff
@@ -3667,6 +3671,10 @@ static function printAjaxPager($title, $start, $numrows, $additional_info = '',
          $back = $start-$list_limit;
       }
 
+      if (!empty($additional_params) && strpos($additional_params, '&') === 0) {
```
Comment (orthagh) [inline inc/html.class.php]: this part is strange.
I read it as:
- if `$additional_param` string is not empty (strlen better anyway) and if char '&' found at position 0 in this string
- then append, at position 0, the char '&'

is the use case should be prefix the string with the char '&' if not already the case ?

File: inc/html.class.php [hunk 2]
```diff
@@ -3923,14 +3931,19 @@ static function printPager($start, $numrows, $target, $parameters, $item_type_ou
...
+    * @param $additional_params  Additional parameters to pass to tab reload request (default '')
     *
     * ajax Pager will be displayed if empty
     *
     * @return void|string
    **/
-   static function printPagerForm($action = "", $display = true) {
+   static function printPagerForm($action = "", $display = true, $additional_params = '') {
+
+      if (!empty($additional_params) && strpos($additional_params, '&') === 0) {
```
Comment (orthagh) [inline inc/html.class.php]: same as previous

File: inc/log.class.php [hunk 3]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
...
+
+      $values = [];
+      while ($data = $iterator->next()) {
+         if (empty($data['user_name'])) {
+            continue;
+         }
+         $values[$data['user_name']] = $data['user_name'];
+      }
+
+      asort($values, SORT_NATURAL | SORT_FLAG_CASE);
```
Comment (orthagh) [inline inc/log.class.php]: why not sorting directly in the query ?

File: inc/dropdown.class.php [hunk 4]
```diff
@@ -1839,7 +1839,7 @@ static function showFromArray($name, array $elements, $options = []) {
 
       $param['option_tooltips'] = Html::entities_deep($param['option_tooltips']);
 
-      if ($param["display_emptychoice"]) {
+      if (!$param['noselect2'] && !$param["multiple"] && $param["display_emptychoice"]) {
```
[the remark this answers is not in the document]
Comment (cedric-anne) [reply inc/dropdown.class.php]: This test was not right, I fixed it in commit a3a67bfd5a273ef06e8928c757047e2de9330b99.
When using select2 for multiple values, adding this empty element makes it selectable. Is is replaced byt the usage of `placeholder` configuration item (https://github.com/glpi-project/glpi/pull/3891/files#diff-527ac4764462540ccf1068f5fa312137R1978).

File: inc/html.class.php (hunk 1 above)
Comment (cedric-anne) [reply inc/html.class.php]: Yes, it was what I expected to do. Fixed in commit 1e9ea81c4ba25225ab270bd90f611fa011a52854.

File: inc/html.class.php (hunk 2 above)
Comment (cedric-anne) [reply inc/html.class.php]: idem

File: inc/log.class.php [hunk 5]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
...
+
+      $values = [];
+      while ($data = $iterator->next()) {
+         $key = $data["linked_action"];
+         $value = null;
+
+         // This is an internal device ?
+         if ($data["linked_action"]) {
+            // Yes it is an internal device
+            switch ($data["linked_action"]) {
```
[the remark this answers is not in the document]
Comment (cedric-anne) [reply inc/log.class.php]: In order to factorize, I think we should use homogeneous articles. Some texts are using the `a` article which can fits both case, where some other are using the `the` article which is made to be dynamic (example: `sprintf(__('%1$s: %2$s'), __('Delete the component'), $data["old_value"])` VS `__('Delete a component')`).

File: inc/log.class.php (hunk 3 above)
Comment (cedric-anne) [reply inc/log.class.php]: I wanted to make a case insensitive natural sort, and I did not found how to do it using native SQL language capabilities.

File: inc/html.class.php [hunk 6]
```diff
@@ -3923,14 +3931,19 @@ static function printPager($start, $numrows, $target, $parameters, $item_type_ou
    /**
     * Display the list_limit combo choice
     *
-    * @param $action    page would be posted when change the value (URL + param) (default '')
-    * @param $display   display the pager form if true, return it if false
+    * @param $action             page would be posted when change the value (URL + param) (default '')
```
Comment (trasher) [inline inc/html.class.php]: Variable type missing beetween param and variable name

File: inc/log.class.php [hunk 7]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
       return $changes;
    }
 
+   /**
+    * Retrieve distinct values for user_name field in item log.
```
Comment (trasher) [inline inc/log.class.php]: missing since

File: inc/log.class.php [hunk 8]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
...
+         $values[$data['user_name']] = $data['user_name'];
+      }
+
+      asort($values, SORT_NATURAL | SORT_FLAG_CASE);
+
+      return $values;
+   }
+
+   /**
+    * Retrieve distinct values for affected field in item log.
```
Comment (trasher) [inline inc/log.class.php]: missing since also

File: inc/log.class.php [hunk 9]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
...
+    * @return array
+    **/
+   static function getDistinctAffectedFieldValuesInItemLog(CommonDBTM $item) {
+      global $DB;
+
+      $itemtype = $item->getType();
+      $items_id = $item->getField('id');
+
+      $iterator = $DB->request([
+         'SELECT DISTINCT' => ['linked_action', 'itemtype_link', 'id_search_option'],
```
Comment (trasher) [inline inc/log.class.php]: not sure distinct will work as expected on several fields

File: inc/log.class.php [hunk 10]
```diff
@@ -704,6 +823,367 @@ static function getHistoryData(CommonDBTM $item, $start = 0, $limit = 0, array $
...
+
+            return strcmp($a, $b);
+         }
+      );
+
+      return $values;
+   }
+
+   /**
+    * Retrieve distinct values for action in item log.
```
Comment (trasher) [inline inc/log.class.php]: missing since also

File: inc/log.class.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (cedric-anne) [reply inc/log.class.php]: Done in commit f34462212c26558b8c52df65266ef7db60458c7e.

File: inc/html.class.php (hunk 6 above)
Comment (cedric-anne) [reply inc/html.class.php]: Fixed in commit f34462212c26558b8c52df65266ef7db60458c7e.

File: inc/log.class.php (hunk 7 above)
Comment (cedric-anne) [reply inc/log.class.php]: Fixed in commit f34462212c26558b8c52df65266ef7db60458c7e.

File: inc/log.class.php (hunk 8 above)
Comment (cedric-anne) [reply inc/log.class.php]: idem

File: inc/log.class.php (hunk 10 above)
Comment (cedric-anne) [reply inc/log.class.php]: idem

File: inc/log.class.php (hunk 9 above)
Comment (cedric-anne) [reply inc/log.class.php]: You are right. I updated in commit f34462212c26558b8c52df65266ef7db60458c7e it to use a `GROUP BY`.

File: inc/html.class.php (hunk 6 above)
Comment (cedric-anne) [reply inc/html.class.php]: Fixed.

**Comment by cedric-anne** [comment]
I fixed it : https://github.com/glpi-project/glpi/pull/3891/files#diff-fe4aeb7ba4bfe5050b46d913b61e60aeR4016

File: inc/dropdown.class.php [hunk 11]
```diff
@@ -1839,7 +1839,7 @@ static function showFromArray($name, array $elements, $options = []) {
 
       $param['option_tooltips'] = Html::entities_deep($param['option_tooltips']);
 
-      if ($param["display_emptychoice"]) {
+      if ($param["display_emptychoice"] && ($param['noselect2'] || !$param["multiple"])) {
```
Comment (orthagh) [inline inc/dropdown.class.php]: ok, last comment for me, we should indicate in dockblock that using `multiple` parameter is incompatible with `display_emptychoice`

Apart this last, PR ok for me

Comment (cedric-anne) [reply inc/dropdown.class.php]: Done.

**Pull request merged**